# COMP90016 - Assignment 2
Version 1 Last edited 10/4/2024

## Semester 1, 2024

In [1]:
NAME = "Devavrath Menon"

ID = "1366694"

This assignment should be completed by each student individually. Make sure you read this entire document, and ask for help if anything is not clear. Any changes or clarifications to this document will be announced via the LMS.

Please make sure you review the University's rules on academic integrity: https://academicintegrity.unimelb.edu.au/

You submission must be your own work. Do not copy material from other students, from the internet or from AI tools. 

Your completed notebook file containing all your answers will be turned in via Canvas. Please also submit an HTML file with the output cleared.

To complete the assignment, finish the tasks in this notebook.

The tasks are a combination of writing your own code, interpreting the results and answering related short-answer questions.

In some cases, we have provided test input and test output that you can use to try out your solutions. These tests are just samples and are **not** exhaustive - they may warn you if you've made a mistake, but they are not guaranteed to. It's up to you to decide whether your code is correct.

**Remember to save your work early and often.**

## Marking

Cells that must be completed to receive marks are clearly labelled. Some cells are code cells, in which you must complete the code to solve a problem. Others are markdown cells, in which you must write your answers to short-answer questions. 

Cells that must be completed to receive marks are labelled like this:

`# -- GRADED CELL (1 mark) - complete this cell --`

Some graded cells are code cells, in which you must complete the code to solve a problem. Other graded cells are markdown cells, in which you must write your answers to short-answer questions. 

You will see the following text in graded code cells:

```
# YOUR CODE HERE
raise NotImplementedError()
```

***You must remove the `raise NotImplementedError()` line from the cell, and replace it with your solution.***

Only add answers to graded cells. If you want to import a library or use a helper function, this must be included in a graded cell.

Only graded cells will be marked.
**Don't make changes outside graded cells, and don't add or remove cells from the notebook**.

>Word limits, where stated, will be strictly enforced. Answers exceeding the limit **will not be marked**.

>Run-time limits will be imposed for each coding question. The run-time of a code cell can be calculated by including `%time` at the top of your cell. Cells exceeding the run-time limit **will not be marked**. The run-time limits only apply to test cases that are included in this document.

No marks are allocated to commenting in your code. We do however, encourage efficient and well-commented code.

The total marks for the assignment add up to 100, and it will be worth 15% of your overall subject grade.

Part 1: 35 marks

Part 2: 20 marks

Part 3: 45 marks


## Submitting

Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and student ID at the top of this notebook.


Your completed notebook file containing all your answers must be turned in via LMS in `.ipynb` format.
You must also submit a copy of this notebook in `html` format with the output cleared.
You can do this by using the `clear all output` option in the menu.

Your submission should include **only two** files with names formatted as: **Assignment_2.ipynb** and **Assignment_2.html**


## Overview

In this assignment, you will answer questions about VCF files and BLAST.

You will use the `pysam` and `biopython` libraries in your functions. You may want to refer to sections of the documentation for these tools for additional help. Additional to `pysam`, `biopython` and standard Python 3 functions and methods, you may also use any other library we have used in Computational Genomics including `collections`, `numpy`, `pandas`, `math`, `itertools`, `seaborn` and `matplotlib`.

You will also use the `BLAST` command-line tool. 

## Setup

In [2]:
import os
import requests
from IPython.core.display import HTML

# Function to get data. DO NOT MODIFY!
def fetch_file(url, outpath='.'):
    response = requests.get(url)
    if response.status_code == 200:
        print('File found!')
        # Get the filename from the URL
        filename = os.path.basename(url).split('?', 1)[0]
        # Construct the filepath using the specified directory and filename
        filepath = os.path.join(outpath, filename)
        # Create the directory if it doesn't exist
        if not os.path.exists(outpath):
            print(f'Creating output dir: {outpath}')
            os.makedirs(outpath)
        # Check if the file already exists in the specified directory
        if os.path.exists(filepath):
            print(f'{filename} already exists in {outpath}. Skip download.')
        else:
            with open(filepath, 'wb') as f:
                f.write(response.content)
                f.close()
            print(f'Saved to: {filepath}')
    else:
        print(f'File not found: Code {response.status_code}')

In [3]:
# Make the notebook pretty
HTML(requests.get('https://raw.githubusercontent.com/melbournebioinformatics/COMP90016/main/data/2023/style/custom.css').text)

In [4]:
# Fetch assignment data
f_names = ["comp90016_assignment_2_demo_a.vcf",
           "comp90016_assignment_2_demo_b.vcf",
           "comp90016_assignment_2.vcf",
           "comp90016_assignment_2_mbp.vcf",
           "comp90016_assignment_2_COVID19_N.fasta",
           "comp90016_assignment_2_9b.fasta"]

for filename in f_names:
    url = f'https://github.com/melbournebioinformatics/COMP90016/blob/main/data/2024/Assignment_02/data/{filename}?raw=true'
    fetch_file(url,'data')

File found!
Saved to: data/comp90016_assignment_2_demo_a.vcf
File found!
Saved to: data/comp90016_assignment_2_demo_b.vcf
File found!
Saved to: data/comp90016_assignment_2.vcf
File found!
Saved to: data/comp90016_assignment_2_mbp.vcf
File found!
Saved to: data/comp90016_assignment_2_COVID19_N.fasta
File found!
Saved to: data/comp90016_assignment_2_9b.fasta


## Part 1: VCF files

### Setup

In [5]:
# Import the pysam module.
import pysam

In [6]:
# Read in the VCF files to produce VariantFile objects.
demo_vcf_a = pysam.VariantFile("data/comp90016_assignment_2_demo_a.vcf")
demo_vcf_b = pysam.VariantFile("data/comp90016_assignment_2_demo_b.vcf")
comp90016_vcf = pysam.VariantFile("data/comp90016_assignment_2.vcf")

#Confirm object type
comp90016_vcf.is_vcf

True

VariantFile objects allow us to access the data within VCF files. You may want to refer back to the work you did in workshop 5. You may also want to read the documentation linked below:
>https://pysam.readthedocs.io/en/latest/usage.html#working-with-vcf-bcf-formatted-files

### Questions
In the cells below, complete the following tasks:

<div class="question">
<h3>Question 1.1</h3>
    
(5 marks, max 50 words)

Which reference genome (full species name and version) was used for the variant calling  in `comp90016_assignment_2.vcf`? Which variant calling tool was used?

</div>


#### -- GRADED CELL (5 marks) - complete this cell -- 

Homo sapiens, 38th version. Freebayes was used.

<div class="info">
<h3> Question 1.2 </h3>

(10 marks)  
    
The four standard DNA bases can be divided into purines (A, G) and pyrimidines (C,T). A transition is a single base change from a purine to a purine, or from a pyrimidine to a pyrimidine. A transversion is a single base change from a purine to a pyrimidine or from a pyrimidine to a purine. The transition/transversion ratio (ti/tv) is the ratio of the number of transitions to the number of transversions.
 
<b>Challenge:</b> Write a Python function to calculate the ti/tv for a set of SNVs in a VCF file containing variants from a single sample.

- [ ] Input: a pysam VariantFile object.
- [ ] Output: Return a single floating-point number. Round to 2 decimal places.
- [ ] Disregard variants that are not SNVs.
- [ ] Consider multi-allelic SNVs as seperate SNVs.
- [ ] Consider a particular SNV allele only once, regardless of how many copies of the allele are present.
- [ ] If there are no transversion SNVs, return None.

</div>

In [7]:
# GRADED CELL 1.2 (10 marks, max 1 min run-time)
%time
def titv(vcf):
    transitions = 0
    transversions = 0

    # Define purines and pyrimidines
    purines = {'A', 'G'}
    pyrimidines = {'C', 'T'}

    # Function to determine if the variant is a transition
    def is_transition(ref, alt):
        if (ref in purines and alt in purines) or (ref in pyrimidines and alt in pyrimidines):
            return True
        return False

    # Function to determine if the variant is a transversion
    def is_transversion(ref, alt):
        if (ref in purines and alt in pyrimidines) or (ref in pyrimidines and alt in purines):
            return True
        return False

    # Process each record in the VCF file
    for rec in vcf.fetch():
        if len(rec.ref) == 1:  # Ensure the reference allele is a single nucleotide
            alt_alleles = set(rec.alts)  # Use set to consider each alternate allele only once
            for alt in alt_alleles:
                if len(alt) == 1 and len(rec.ref) == 1:  # Ensure both ref and alt are single nucleotides and the same length
                    if is_transition(rec.ref, alt):
                        transitions += 1
                    elif is_transversion(rec.ref, alt):
                        transversions += 1

    # Handle case where there are no transversions
    if transversions == 0:
        return None

    return transitions / transversions

        



CPU times: user 2 µs, sys: 0 ns, total: 2 µs
Wall time: 7.15 µs


In [8]:
# Test your function in this cell
print(titv(demo_vcf_a)) # should output 2.0
print(titv(demo_vcf_b)) # should output 1.0

print(titv(comp90016_vcf))

2.0
1.0
2.3214285714285716


In [9]:
# --- AUTOGRADING CELL DO NOT EDIT ----


<div class="question">
<h3>Question 1.3</h3>
    
(10 marks, max 100 words)

Explain why the ti/tv ratio is typically higher in protein-coding regions than in non-coding regions.

</div>


#### -- GRADED CELL (5 marks) - complete this cell -- 

The ti/tv ratio is typically higher in protein-coding regions than in non-coding regions due to the nature of DNA mutations and their impact on protein function. Transitions are less likely to alter the amino acid sequence of proteins due to the redundancy of the genetic code so they are more tolerated during evolution. Transversions are more likely to cause disruptive changes in the amino acid sequence, leading to detrimental effects on protein function and lower survival.








<div class="question">
<h3>Question 1.4</h3>
    
(5 marks, max 100 words)

When working on a variant calling project, you might generate files with the extensions vcf.gz or vcf.tbi. Explain the contents of these files and how they relate to .vcf files.

</div>


#### -- GRADED CELL (5 marks) - complete this cell -- 
.vcf files store variant information (chrom, pos, id, ref,alt,qualt) vcf.gz is compressed version of .vcf and vcf.tbi stores indexes. 

<div class="question">
<h3>Question 1.5</h3>
    
(5 marks, max 100 words)

Computational genomics often involves working with large data sets. Explain how .gz and .tbi files help make working with large files feasible.

</div>


#### -- GRADED CELL (5 marks) - complete this cell -- 

gz allows for storing the data by reducing size through compressing the large data in .vcf.

.tbi that stores indexes allows for accessing the information from .gz without the need to decompress the file. 

Overall effect is space efficiency, faster sharing and analysis. 

## Part 2: More VCF files

### Setup

The data you will be using are publically available as part of the 1000 Genomes Project, which was completed in 2015. It was one of the first large-scale projects to try and capture the level of variation in the human genome. You can read one of the key publiciations describing this dataset here: https://www.nature.com/articles/nature15393. 

Our gene of interest is *MBP* (Myelin Basic Protein, NM_001025081). Variants in *MBP* from three selected individuals (HG02687, HG03897 and HG04118) have been included.

In [10]:
# Read in the MBP VCF file to produce a VariantFile objects.
comp90016_vcf_MBP = pysam.VariantFile("data/comp90016_assignment_2_mbp.vcf")

### Questions
In the cells below, complete the following tasks:

<div class="info">
<h3> Question 2.1 </h3>

(10 marks)  
 
<b>Challenge:</b> Write a Python function to create a dictionary of for a given variant position across all samples in a pysam VariantFile object.

- [ ] Input:
    - a pysam VariantFile object (vcf)
    - the name of a contig as a string (chrom)
    - an integer describing a position (pos)
- [ ] Output: A dictionary of genotypes with sample names as keys and tuples of the genotypes as values.
- [ ] If the combination of chrom and pos does not exist in vcf, return None.
    
</div>

In [11]:
# GRADED CELL 2.1 (10 marks, max 1 min run-time)

%time
def genotype_dict(vcf, chrom, pos):
    answerdic = {}   #placeholder 
    verifchrom = None
    verifpos = None
    for rec in vcf.fetch(): 
        if str(rec.contig) == chrom and int(rec.pos) == pos:
            sample_list = list(rec.header.samples)
            for sample in sample_list:
                answerdic[sample] = rec.samples[sample]['GT']
    if len(answerdic) != 0:
        return answerdic
    else:
        return None
     



CPU times: user 4 µs, sys: 1 µs, total: 5 µs
Wall time: 16.9 µs


In [12]:
# Test your function in this cell
print(genotype_dict(demo_vcf_b, '18', 74690979)) # should output {'HG02687': (0, 0), 'HG03897': (0, 1), 'HG04118': (0, 0)}
print(genotype_dict(demo_vcf_b, '2', 1000)) # should output None
print(genotype_dict(comp90016_vcf_MBP, '18', 74694135))


{'HG02687': (0, 0), 'HG03897': (0, 1), 'HG04118': (0, 0)}
None
{'HG02687': (1, 1), 'HG03897': (1, 0), 'HG04118': (1, 1)}


In [13]:
# --- AUTOGRADING CELL DO NOT EDIT ----


<div class="question">
<h3>Question 2.2</h3>
    
(5 marks, max 100 words)

For each sample, list the total number of homozygous reference, homozygous alternative and heterozygous alleles. Which of the three samples has the most homozygous alternative variants in the MBP gene?
</div>


#### -- GRADED CELL (5 marks) - complete this cell -- 
HG02687: Homozygous Reference = 1317, Heterozygous = 68, Homozygous Alternative = 22 


HG03897: Homozygous Reference = 1349, Heterozygous = 42, Homozygous Alternative = 16


HG04118: Homozygous Reference = 1331, Heterozygous = 58, Homozygous Alternative = 18



Among these samples, HG02687 has the most homozygous alternative variants in the MBP gene, with a total of 22.

<div class="question">
<h3>Question 2.3</h3>
    
(5 marks, max 100 words)

What information is contained in the ID column of the MBP VCF file? Explain why this information is useful.
    
</div>


#### -- GRADED CELL (5 marks) - complete this cell -- 

If there is a record of that variant in some database the ID column keeps the identifier that we can use to access the record in database. This can make it easier to cross reference, check whether the variant is associated with any traits, diseases. 

## Part 3: BLAST

### Setup

The gene ORF-9b has been detected in multiple isolates of SARS coronavirus as well as in other coronaviruses that infect non-human animals. We will determine whether SARS-CoV-2 (responsible for COVID-19) has an ORF that is homologous to ORF-9b in other coronaviruses. We will be using `BLAST` on the command line for this purpose.

We will be using a sequence from an isolate of SARS-CoV-2 sequenced early in the COVID-19 pandemic, in December 2019 in Wuhan, China. `data/comp90016_assignment_2_COVID19_N.fasta` is a FASTA file of the DNA sequence for the nucleocapsid (N) gene.

`data/comp90016_assignment_2_9b.fasta` contains ORF-9b sequences from a selection of characterised coronaviruses.

We will be using command line BLAST to execute a search.

BLAST can be downloaded [here](https://www.ncbi.nlm.nih.gov/books/NBK279671/) OR installed using conda `conda install -c bioconda blast`

Execute the following code block to complete the search. Note that `%%bash` signifies that the code block contains Linux commands and not Python code.

In [14]:
# Import Biopython to take advantage of Bio.Align objects for storing and accessing sequence alignments.
from Bio import Align, Seq, SeqRecord
# You may want to load the BLAST results with BlastIO https://biopython.org/docs/1.75/api/Bio.SearchIO.BlastIO.html

In [15]:
%%bash
# First let's check that BLAST is installed in the current environment and is available on your PATH
which blastn
# If the output is blank or you see an error, BLAST is not properly installed

/srv/conda/envs/notebook/bin/blastn


In [16]:
%%bash

# Build a custom BLAST database from the ORF-9b sequences.
echo "Prepare blast database"
makeblastdb -dbtype nucl -in data/comp90016_assignment_2_9b.fasta

# You will see some extra files have appeared.

# Execute a BLAST search.
echo "Running blastn"
# The results will be returned in tabulated form and saved to a file.
blastn -query data/comp90016_assignment_2_COVID19_N.fasta -db data/comp90016_assignment_2_9b.fasta -outfmt 7 -out blast_results.tab

Prepare blast database


Building a new DB, current time: 04/21/2024 04:24:09
New DB name:   /home/jovyan/data/comp90016_assignment_2_9b.fasta
New DB title:  data/comp90016_assignment_2_9b.fasta
Sequence type: Nucleotide
Keep MBits: T
Maximum file size: 3000000000B
Adding sequences from FASTA; added 6 sequences in 0.00126696 seconds.


Running blastn


In [17]:
%%bash
# Print contents of blastn output
cat blast_results.tab

# BLASTN 2.14.1+
# Query: GeneID:43740575|locus:GU280_gp10|Symbol:N|GeneProductName:nucleocapsid phosphoprotein|Organism:Severe acute respiratory syndrome coronavirus 2
# Database: data/comp90016_assignment_2_9b.fasta
# Fields: query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score
# 6 hits found
GeneID:43740575|locus:GU280_gp10|Symbol:N|GeneProductName:nucleocapsid	bat_coronavirus_HKU3_orf_9b	89.116	294	32	0	11	304	1	294	7.49e-105	366
GeneID:43740575|locus:GU280_gp10|Symbol:N|GeneProductName:nucleocapsid	human_SARS_coronavirus_Tor2_orf_9b	88.552	297	31	2	11	304	1	297	4.51e-102	357
GeneID:43740575|locus:GU280_gp10|Symbol:N|GeneProductName:nucleocapsid	human_SARS_coronavirus_LLJ_orf_9b	88.552	297	31	2	11	304	1	297	4.51e-102	357
GeneID:43740575|locus:GU280_gp10|Symbol:N|GeneProductName:nucleocapsid	human_SARS_coronavirus_ZJ01_orf_9b	88.552	297	31	2	11	304	1	297	4.51e-102	357
GeneID:43740575|locus:GU280_gp

### Questions
In the cells below, complete the following tasks:

<div class="question">

<h3> Question 3.1 </h3>

(5 marks, max 100 words)

BLAST can be used to query a database for matches to a sequence of interest. Here we have created a custom database from a selection of coronavirus genomes, but we may also want to search for matches in a public database such as those available via NCBI, or in a custom database shared by another researcher.
    

Explain one factor that is important to consider when selecting an appropriate BLAST database.
    
</div>



#### -- GRADED CELL (5 marks) - complete this cell -- 
Accuracy required, expert curated database are more likely to be accurate than public uncurated databases. Using a database tailored to your research scope, whether it's focused on specific organisms, diseases, or genetic regions, enhances the likelihood of discovering biologically and clinically relevant matches.

<div class="question">

<h3> Question 3.2 </h3>

(10 marks, max 100 words)

Is there evidence from the BLAST results that the SARS-CoV-2 N gene has an internal ORF homologous to ORF-9b sequences in other coronaviruses? Justify your choice. What was the highest scoring hit?
    
</div>



#### -- GRADED CELL (10 marks) - complete this cell -- 

6 hits were found with bat_coronavirus_HKU3_orf_9b5 having the highest bit score of 366, strong E-value. While its percentage identity is slightly less than bat_coronavirus_RM1_orf_9b with 89.151, I choose bat_coronavirus_HKU3_orf_9b5 because of the strong bit score, higher alignment length and  E-value. 

<div class="question">

<h3> Question 3.3 </h3>

(5 marks, max 100 words)

Interpret the E value of the top scoring BLAST hit. How confident are you in your BLAST results?
    
</div>



#### -- GRADED CELL (5 marks) - complete this cell -- 
I am very confident in 7.49e-105, since this a really small number and smaller the E-value, higher the confidence. Unlikely to be by chance.

<div class="question">

<h3> Question 3.4 </h3>

(5 marks, max 50 words)

Modify the previous BLAST command to:

- Work with an input of a protein sequence FASTA file (named `comp90016_assignment_2_COVID19_N_protein.fasta`). 
- Decrease the word-size parameter from the default value to 2. 
- Use the same custom database and output format used previously. 
    
**Note that you are not required to execute this command.**
    
</div>



#### -- GRADED CELL (5 marks) - complete this cell -- 
Reducing the word count increases the sensitivity, has more hits and  thus takes longer as it processes those reads. 

tblastn -query comp90016_assignment_2_COVID19_N_protein.fasta -db data/comp90016_assignment_2_9b.fasta -outfmt 7 -out blast_results.tab -word_size 2



<div class="question">

<h3> Question 3.5 </h3>

(10 marks, max 100 words)

What effect does decreasing the word size have on the sensitivity (ability to detect true homologs) and run-time of the search? Explain why this is the case.
    
</div>



#### -- GRADED CELL (10 marks) - complete this cell -- 

Decreasing the word size in a BLAST search increases the sensitivity of the search, allowing it to detect more true homologs, including those with lower similarity. This is because a smaller word size lowers the threshold for initial matches, thus picking up more potential alignments. However, this also increases the run-time since the algorithm needs to process and extend more initial hits. 

For the next question, we will be using amino-acid substitution matrices, encoded as 2D dictionaries. The cell below stores the BLOSUM-62 substitution matrix as a dictionary (the default substitution matrix for BLAST). The protein sequences used to test your code will only contain the one-letter codes of the 20 standard amino-acids.

In [18]:
BLOSUM_62 = {
    '*':{'*':1,'A':-4,'C':-4,'B':-4,'E':-4,'D':-4,'G':-4,'F':-4,'I':-4,'H':-4,'K':-4,'M':-4,'L':-4,'N':-4,'Q':-4,'P':-4,'S':-4,'R':-4,'T':-4,'W':-4,'V':-4,'Y':-4,'X':-4,'Z':-4},
    'A':{'*':-4,'A':4,'C':0,'B':-2,'E':-1,'D':-2,'G':0,'F':-2,'I':-1,'H':-2,'K':-1,'M':-1,'L':-1,'N':-2,'Q':-1,'P':-1,'S':1,'R':-1,'T':0,'W':-3,'V':0,'Y':-2,'X':-1,'Z':-1},
    'C':{'*':-4,'A':0,'C':9,'B':-3,'E':-4,'D':-3,'G':-3,'F':-2,'I':-1,'H':-3,'K':-3,'M':-1,'L':-1,'N':-3,'Q':-3,'P':-3,'S':-1,'R':-3,'T':-1,'W':-2,'V':-1,'Y':-2,'X':-1,'Z':-3},
    'B':{'*':-4,'A':-2,'C':-3,'B':4,'E':1,'D':4,'G':-1,'F':-3,'I':-3,'H':0,'K':0,'M':-3,'L':-4,'N':3,'Q':0,'P':-2,'S':0,'R':-1,'T':-1,'W':-4,'V':-3,'Y':-3,'X':-1,'Z':1},
    'E':{'*':-4,'A':-1,'C':-4,'B':1,'E':5,'D':2,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':-2,'L':-3,'N':0,'Q':2,'P':-1,'S':0,'R':0,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':4},
    'D':{'*':-4,'A':-2,'C':-3,'B':4,'E':2,'D':6,'G':-1,'F':-3,'I':-3,'H':-1,'K':-1,'M':-3,'L':-4,'N':1,'Q':0,'P':-1,'S':0,'R':-2,'T':-1,'W':-4,'V':-3,'Y':-3,'X':-1,'Z':1},
    'G':{'*':-4,'A':0,'C':-3,'B':-1,'E':-2,'D':-1,'G':6,'F':-3,'I':-4,'H':-2,'K':-2,'M':-3,'L':-4,'N':0,'Q':-2,'P':-2,'S':0,'R':-2,'T':-2,'W':-2,'V':-3,'Y':-3,'X':-1,'Z':-2},
    'F':{'*':-4,'A':-2,'C':-2,'B':-3,'E':-3,'D':-3,'G':-3,'F':6,'I':0,'H':-1,'K':-3,'M':0,'L':0,'N':-3,'Q':-3,'P':-4,'S':-2,'R':-3,'T':-2,'W':1,'V':-1,'Y':3,'X':-1,'Z':-3},
    'I':{'*':-4,'A':-1,'C':-1,'B':-3,'E':-3,'D':-3,'G':-4,'F':0,'I':4,'H':-3,'K':-3,'M':1,'L':2,'N':-3,'Q':-3,'P':-3,'S':-2,'R':-3,'T':-1,'W':-3,'V':3,'Y':-1,'X':-1,'Z':-3},
    'H':{'*':-4,'A':-2,'C':-3,'B':0,'E':0,'D':-1,'G':-2,'F':-1,'I':-3,'H':8,'K':-1,'M':-2,'L':-3,'N':1,'Q':0,'P':-2,'S':-1,'R':0,'T':-2,'W':-2,'V':-3,'Y':2,'X':-1,'Z':0},
    'K':{'*':-4,'A':-1,'C':-3,'B':0,'E':1,'D':-1,'G':-2,'F':-3,'I':-3,'H':-1,'K':5,'M':-1,'L':-2,'N':0,'Q':1,'P':-1,'S':0,'R':2,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':1},
    'M':{'*':-4,'A':-1,'C':-1,'B':-3,'E':-2,'D':-3,'G':-3,'F':0,'I':1,'H':-2,'K':-1,'M':5,'L':2,'N':-2,'Q':0,'P':-2,'S':-1,'R':-1,'T':-1,'W':-1,'V':1,'Y':-1,'X':-1,'Z':-1},
    'L':{'*':-4,'A':-1,'C':-1,'B':-4,'E':-3,'D':-4,'G':-4,'F':0,'I':2,'H':-3,'K':-2,'M':2,'L':4,'N':-3,'Q':-2,'P':-3,'S':-2,'R':-2,'T':-1,'W':-2,'V':1,'Y':-1,'X':-1,'Z':-3},
    'N':{'*':-4,'A':-2,'C':-3,'B':3,'E':0,'D':1,'G':0,'F':-3,'I':-3,'H':1,'K':0,'M':-2,'L':-3,'N':6,'Q':0,'P':-2,'S':1,'R':0,'T':0,'W':-4,'V':-3,'Y':-2,'X':-1,'Z':0},
    'Q':{'*':-4,'A':-1,'C':-3,'B':0,'E':2,'D':0,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':0,'L':-2,'N':0,'Q':5,'P':-1,'S':0,'R':1,'T':-1,'W':-2,'V':-2,'Y':-1,'X':-1,'Z':3},
    'P':{'*':-4,'A':-1,'C':-3,'B':-2,'E':-1,'D':-1,'G':-2,'F':-4,'I':-3,'H':-2,'K':-1,'M':-2,'L':-3,'N':-2,'Q':-1,'P':7,'S':-1,'R':-2,'T':-1,'W':-4,'V':-2,'Y':-3,'X':-1,'Z':-1},
    'S':{'*':-4,'A':1,'C':-1,'B':0,'E':0,'D':0,'G':0,'F':-2,'I':-2,'H':-1,'K':0,'M':-1,'L':-2,'N':1,'Q':0,'P':-1,'S':4,'R':-1,'T':1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':0},
    'R':{'*':-4,'A':-1,'C':-3,'B':-1,'E':0,'D':-2,'G':-2,'F':-3,'I':-3,'H':0,'K':2,'M':-1,'L':-2,'N':0,'Q':1,'P':-2,'S':-1,'R':5,'T':-1,'W':-3,'V':-3,'Y':-2,'X':-1,'Z':0},
    'T':{'*':-4,'A':0,'C':-1,'B':-1,'E':-1,'D':-1,'G':-2,'F':-2,'I':-1,'H':-2,'K':-1,'M':-1,'L':-1,'N':0,'Q':-1,'P':-1,'S':1,'R':-1,'T':5,'W':-2,'V':0,'Y':-2,'X':-1,'Z':-1},
    'W':{'*':-4,'A':-3,'C':-2,'B':-4,'E':-3,'D':-4,'G':-2,'F':1,'I':-3,'H':-2,'K':-3,'M':-1,'L':-2,'N':-4,'Q':-2,'P':-4,'S':-3,'R':-3,'T':-2,'W':11,'V':-3,'Y':2,'X':-1,'Z':-3},
    'V':{'*':-4,'A':0,'C':-1,'B':-3,'E':-2,'D':-3,'G':-3,'F':-1,'I':3,'H':-3,'K':-2,'M':1,'L':1,'N':-3,'Q':-2,'P':-2,'S':-2,'R':-3,'T':0,'W':-3,'V':4,'Y':-1,'X':-1,'Z':-2},
    'Y':{'*':-4,'A':-2,'C':-2,'B':-3,'E':-2,'D':-3,'G':-3,'F':3,'I':-1,'H':2,'K':-2,'M':-1,'L':-1,'N':-2,'Q':-1,'P':-3,'S':-2,'R':-2,'T':-2,'W':2,'V':-1,'Y':7,'X':-1,'Z':-2},
    'X':{'*':-4,'A':-1,'C':-1,'B':-1,'E':-1,'D':-1,'G':-1,'F':-1,'I':-1,'H':-1,'K':-1,'M':-1,'L':-1,'N':-1,'Q':-1,'P':-1,'S':-1,'R':-1,'T':-1,'W':-1,'V':-1,'Y':-1,'X':-1,'Z':-1},
    'Z':{'*':-4,'A':-1,'C':-3,'B':1,'E':4,'D':1,'G':-2,'F':-3,'I':-3,'H':0,'K':1,'M':-1,'L':-3,'N':0,'Q':3,'P':-1,'S':0,'R':0,'T':-1,'W':-3,'V':-2,'Y':-2,'X':-1,'Z':4}}

<div class="info">
<h3> Question 3.6 </h3>

(10 marks)

<b>Challenge:</b> Write a Python function to calculate the percentage sequence similarity between two aligned protein sequences. The percentage similarity is defined as the number of positions with similar amino acid residues as a percentage of the total number of positions in the alignment. 
    
Two amino acid residues are considered similar based on the input substitution matrix. If two amino acids have a value greater than 0 in their corresponding cell of the input matrix, they are similar. 

- [ ] Input: 
    - Pairwise_aln as a `Bio.Align.MultipleSeqAlignment` object that contains only two aligned sequences.
    - sub_matrix as a 2D dictionary encoding an amino-acid substitution matrix, such as the BLOSUM-62 matrix above.
- [ ] Output: Return a floating-point number between 0 and 100. 
- [ ] Round the output to 4 decimal places.
- [ ] If pairwise_aln does not contain 2 sequences, return None.

</div>

In [19]:
# GRADED CELL 3.6 (10 marks, max 1 min run-time)
%time
def percentage_similarity(pairwise_aln, sub_matrix):
    if len(pairwise_aln) != 2:
        return None
    first_seq, second_seq = pairwise_aln[0], pairwise_aln[1]
    position_count = len(first_seq)

    similar_count = 0  # total number of similar positions.

    for pos_index in range(position_count): 
        if first_seq[pos_index] == '-' or second_seq[pos_index] == '-':  # to handle case for missing AA.
            continue
        try:
            if sub_matrix[first_seq[pos_index]][second_seq[pos_index]] > 0:
                similar_count += 1
        except KeyError: 
            print(f"Invalid amino acid pair: {first_seq[pos_index]}, {second_seq[pos_index]}")

    return round(similar_count / position_count * 100, 4)


CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 7.63 µs


In [20]:
# ~~ Test your function in this cell ~~
demo_sequence_a = SeqRecord.SeqRecord(Seq.Seq("MYWIW"))
demo_sequence_b = SeqRecord.SeqRecord(Seq.Seq("IYW--"))
demo_pairwise_prot_aln = Align.MultipleSeqAlignment([demo_sequence_a, demo_sequence_b])

sequence_a = SeqRecord.SeqRecord(Seq.Seq("MYGEGEPGGWQDHVTVLATRRHPKWAQAWVSTMPWGYECGFSRAWVHQTPWINV-----VSLSSHEAYGVVAVRHPWEIFSPYEVYAPYVQDTQHHGNPGQFTTSCYPDE"))
sequence_b = SeqRecord.SeqRecord(Seq.Seq("MYADGEPGAWQDHMTVLAIYWHHKWAHAWVSTMPWSYECGFSRAWVHQTPWINVIRFTQVSLSSRAWYGILAVRHPWEIFSPYDVYAPYVAATQHHGNPGQFSTSCYP--"))
pairwise_prot_aln = Align.MultipleSeqAlignment([sequence_a, sequence_b])

print(percentage_similarity(demo_pairwise_prot_aln, BLOSUM_62)) # should return 60.0

print(percentage_similarity(pairwise_prot_aln, BLOSUM_62))

60.0
81.8182


In [21]:
# --- AUTOGRADING CELL DO NOT EDIT ----


# END OF ASSIGNMENT

## Submitting

Follow these steps to submit your assignment

1) Before you turn this assignment in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

2) Make sure you have filled in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE"

3) **Clear all output** (in the menubar, select Kernel$\rightarrow$Restart & Clear Output)

4) Your completed notebook file containing all your answers must be turned in via Canvas in `.ipynb` format.

5) You must also submit a copy of this notebook in `html` format with the output cleared.


Your submission should include **only two** files with names formatted as: **Assignment_2.ipynb** and **Assignment_2.html**
